edgexfoundry/app-rfid-llrp-inventory
# build(snap): update snap packaging

- Update snap base to core 20
- Update documentation
- Update metadata and versioning
- Add support for "snap set" of device-specific settings
- Update install hook
- Other minor updates

Signed-off-by: Siggi Skulason <siggi.skulason@canonical.com>

Fix https://github.com/edgexfoundry/app-rfid-llrp-inventory/issues/58







**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/app-rfid-llrp-inventory/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

  <link to docs PR>

## Testing Instructions


It's not possible to fully test this without a reader. What I did, for reference, is:

1. Remove and reinstall edgexfoundry and the LLRP-RFID snaps

```
sudo snap remove edgexfoundry
sudo snap remove edgex-device-rfid-llrp
sudo snap remove edgex-app-rfid-llrp-inventory
sudo snap install edgexfoundry --edge
```

2. Run the device-rfid-llrp snap from https://github.com/edgexfoundry/device-rfid-llrp-go/pull/61. Connect the interfaces and run the scan to get the subnet information.

```
snapcraft clean --use-lxd
snapcraft --use-lxd
sudo snap install edgex-device-rfid-llrp_1.0.1-dev.7_amd64.snap --dangerous
sudo snap connect edgexfoundry:edgex-secretstore-token edgex-device-rfid-llrp:edgex-secretstore-token
sudo snap start edgex-device-rfid-llrp.device-rfid-llrp 
```

```
sudo snap connect edgex-device-rfid-llrp:network-control 
sudo snap connect edgex-device-rfid-llrp:network-observe 

CONSUL_TOKEN=$(sudo cat /var/snap/edgexfoundry/current/secrets/consul-acl-token/bootstrap_token.json | jq ".SecretID" | tr -d '"') 
echo $CONSUL_TOKEN
sudo edgex-device-rfid-llrp.auto-configure $CONSUL_TOKEN
```

3. Wait for the RFID reader to be detected (takes a few seconds). Confirm using `edgex-cli` that the device was added:

```
$ edgex-cli device list
Name                Description       ServiceName       ProfileName          Labels                           AutoEvents
SpeedwayR-10-EE-EF  LLRP RFID Reader  device-rfid-llrp  LLRP-Impinj-Profile  [RFID LLRP Impinj SpeedwayR420]  []
```

3. Using the name of the device, update the configuration.toml file for this inventory snap, adding the required alias:

```
[AppCustom.Aliases]
SpeedwayR-10-EE-EF_1="office"
```
Note, that as per the Readme, the key is the device name plus "_1" for the 1st antenna.

4. Build, install and run the snap

```
snapcraft --use-lxd
sudo snap install edgex-app-rfid-llrp-inventory_1.0.1-dev.6_amd64.snap --dangerous
sudo snap connect edgexfoundry:edgex-secretstore-token edgex-app-rfid-llrp-inventory:edgex-secretstore-token
sudo snap start edgex-app-rfid-llrp-inventory.app-rfid-llrp-inventory 
```

5. Use the Consul UI to double-check that the AppCustom/Aliases value is correctly set


6. Run a scan for a few seconds and check the results

```
curl -o- -X POST localhost:59711/api/v2/command/reading/start
sleep 5s
curl -o- -X POST localhost:59711/api/v2/command/reading/stop
sleep 3s
curl -o- localhost:59711/api/v2/inventory/snapshot
```
Result:

```
[{"epc":"e200001b690901820340ee3f","tid":"","location":{"device_name":"SpeedwayR-10-EE-EF","antenna_id":1},"location_alias":"office","last_read":1645042805425,"last_arrived":1645042800663,"last_departed":0,"state":"Present","stats_map":{"SpeedwayR-10-EE-EF_1":{"last_read":1645042805425,"mean_rssi":-45.425}}},{"epc":"e200001b690901820330ef6c","tid":"","location":{"device_name":"SpeedwayR-10-EE-EF","antenna_id":1},"location_alias":"office","last_read":1645042805425,"last_arrived":1645042800663,"last_departed":0,"state":"Present","stats_map":{"SpeedwayR-10-EE-EF_1":{"last_read":1645042805425,"mean_rssi":-33.525}}}]
```
 









## New Dependency Instructions (If applicable)
